include/deal.II/base/mpi_tags.h

```diff
@@ -0,0 +1,99 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2019 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii_mpi_tags_h
+#define dealii_mpi_tags_h
+
+#include <deal.II/base/config.h>
+
+DEAL_II_NAMESPACE_OPEN
+
+
+namespace Utilities
+{
+  namespace MPI
+  {
+    namespace internal
+    {
+      /**
+       * This enum holds all MPI tags used in collective MPI communications with
+       * using MPI_ANY_SOURCE inside the deal.II library.
+       *
+       * We keep these tags in a central location so that they are unique within
+       * the library. Otherwise, communication that receives packages with
+       * MPI_ANY_SOURCE might pick up packets from a different algorithm.
+       */
+      struct Tags
+      {
+        /**
+         * The enum with the tags.
+         */
+        enum enumeration
```

Comment:
```
struct Tags {
  enum enumeration {
```
is now exactly equivalent to `enum class Tags`. The latter also lets you choose the type: `enum class Tags : int`.

Reply:
If I use an enum class, I can not convert an enum value to an int without a cast. I think it is quite ugly to have these casts everywhere in the code.

Reply:
Isn't this working:
```cpp
enum Tags : int
{
a = 10,
b,
};

int main()
{
    std::cout << Tags::a << std::endl;
    std::cout << Tags::b << std::endl;
}
```

I am not very familiar with enums...

Reply:
see https://godbolt.org/z/Qkb9he

Reply:
What about this: https://godbolt.org/z/Q9uxbS

Reply:
The ``: int`` won't make a difference as this is the default, so the only difference is that you remove the struct.
If you use ``enum Tags`` the symbols are also available in the scope around it (so in your example ``b`` is now also valid). I put the enum into the structure to force usage as:
``internal::Tags::b`` (otherwise ``internal::b`` would also compile).

Reply:
I see the issue with the conversion. That's awkward. Then just change the `struct` to a `namespace` -- there's no reason anyone would ever want to create an object of that kind.

I'd still stick with `: int` in the `enum`, even just so that you get an error if for some reason we ever overflow the range. In fact, it may be useful to derive from `uint_16t` to ensure we don't step on the "special" tag values.